crates/ruff_linter/src/rules/ruff/rules/helpers.rs

```diff
@@ -84,3 +87,72 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
         })
     })
 }
+
+// fast check to disqualify any string literal that doesn't
+// have opening and closing brackets
+#[inline]
+fn has_brackets(possible_fstring: &str) -> bool {
+    let mut opening = false;
+    let mut closing = false;
+    for c in possible_fstring.chars() {
+        match c {
+            '{' => opening = true,
+            '}' => closing = true,
+            _ => {}
+        }
+    }
+    opening && closing
+}
+
+/// Returns `true` if `source` is valid f-string syntax with qualified, bound variables.
+/// `kwargs` should be the keyword arguments that were passed to function if the string literal is also
+/// being passed to the same function.
+/// If a identifier from `kwargs` is used in `source`'s formatting, this will return `false`,
+/// since it's possible the function could be formatting the literal in question.
+pub(super) fn should_be_fstring(
+    source: &str,
+    kwargs: Option<&Vec<Keyword>>,
+    semantic: &SemanticModel,
+) -> bool {
+    if !has_brackets(source) {
+        return false;
+    }
+
+    let Ok(Expr::FString(ast::ExprFString { value, .. })) =
+        parse_expression(&format!("f\"{source}\""))
+    else {
+        return false;
+    };
+
+    let kw_idents: FxHashSet<String> = kwargs
+        .map(|keywords| {
+            keywords
+                .iter()
+                .filter_map(|k| k.arg.clone())
+                .map(Into::into)
+                .collect()
+        })
+        .unwrap_or_default();
```

Comment:
I believe this can be:

```rust
let kw_idents: FxHashSet<&str> = kwargs
    .map(|keywords| {
        keywords
            .iter()
            .filter_map(|k| k.arg.as_ref())
            .map(|arg| arg.as_str())
            .collect()
    })
    .unwrap_or_default();
```

And then below, `kw_idents.contains(id.as_str())`.